File: backend/internal/model/runlog.go
package model

import "time"

// RunLogStatus indicates the outcome of a logged run.
type RunLogStatus string

const (
	RunLogRunning RunLogStatus = "running"
	RunLogSuccess RunLogStatus = "success"
	RunLogFailed  RunLogStatus = "failed"
)

// ArtifactRef points to a file produced by a run.
type ArtifactRef struct {
	Name string `json:"name"` // e.g. "mock.html", "vision.md"
	Path string `json:"path"` // relative to project HostDir
}

// StepLogRef records a single agent step's trace file within a run directory.
type StepLogRef struct {
	Step     string `json:"step"`               // e.g. "screen_planner", "component_planner_dashboard"
	File     string `json:"file"`               // filename within the run dir, e.g. "screen_planner.txt"
	Status   string `json:"status"`             // "ok" or "error"
	Detail   string `json:"detail,omitempty"`   // short parse/error summary
	Duration int64  `json:"durationMs,omitempty"`
}

// RunLogEntry records the full lifecycle of a single pipeline operation.
// Persisted at <logBase>/<project-name>/runs/<run-id>/meta.json.
type RunLogEntry struct {
	ID           string        `json:"id"`
	ProjectID    string        `json:"projectId"`
	ProjectName  string        `json:"projectName"`
	Stage        StageName     `json:"stage"`
	Operation    string        `json:"operation"`  // chat | mock | beads-generate | beads-execute | summary
	Attempt      int           `json:"attempt"`    // 1-based retry count for this project+stage+op
	Status       RunLogStatus  `json:"status"`
	StartedAt    time.Time     `json:"startedAt"`
	EndedAt      *time.Time    `json:"endedAt,omitempty"`
	DurationMs   int64         `json:"durationMs,omitempty"`
	Artifacts    []ArtifactRef `json:"artifacts,omitempty"`
	Error        string        `json:"error,omitempty"`
	ErrorLogPath string        `json:"errorLogPath,omitempty"` // relative path within run dir
	TokensTotal  int           `json:"tokensTotal,omitempty"`
	Notes        string        `json:"notes,omitempty"` // e.g. "3 epics, 12 tasks created"
	StepLogs     []StepLogRef  `json:"stepLogs,omitempty"`
}

File: frontend/src/api/runlog.ts
import { apiFetch } from './client';

export interface ArtifactRef {
  name: string;
  path: string;
}

export interface StepLogRef {
  step: string;
  file: string;
  status: string;
  detail?: string;
  durationMs?: number;
}

export interface RunLogEntry {
  id: string;
  projectId: string;
  projectName: string;
  stage: string;
  operation: string;
  attempt: number;
  status: 'running' | 'success' | 'failed';
  startedAt: string;
  endedAt?: string;
  durationMs?: number;
  artifacts?: ArtifactRef[];
  error?: string;
  errorLogPath?: string;
  tokensTotal?: number;
  notes?: string;
  stepLogs?: StepLogRef[];
}

export function getProjectRunLog(projectId: string): Promise<RunLogEntry[]> {
  return apiFetch<RunLogEntry[]>(`/projects/${projectId}/run-log`);
}

export function getAllRunLog(): Promise<RunLogEntry[]> {
  return apiFetch<RunLogEntry[]>(`/run-log`);
}

export async function getTraceFile(projectId: string, runId: string, filename: string): Promise<string> {
  const res = await fetch(`/api/projects/${projectId}/run-log/${runId}/trace/${encodeURIComponent(filename)}`, {
    credentials: 'include',
  });
  if (!res.ok) throw new Error(await res.text() || res.statusText);
  return res.text();
}

export function deleteRun(projectId: string, runId: string): Promise<void> {
  return apiFetch<void>(`/projects/${projectId}/run-log/${runId}`, { method: 'DELETE' });
}

export function pruneRuns(projectId: string): Promise<{ deleted: number }> {
  return apiFetch<{ deleted: number }>(`/projects/${projectId}/run-log/prune`, { method: 'POST' });
}
